frame/mixnet/src/lib.rs

```diff
@@ -0,0 +1,762 @@
+				"Initial mixnodes already set"
+			);
+			for (i, mixnode) in self.mixnodes.iter().enumerate() {
+				// We just make up authority indices here. This doesn't matter as authority indices
+				// are only used during registration to check an authority doesn't register twice.
+				EvenSessionMixnodes::<T>::insert(i as AuthorityIndex, mixnode);
+				OddSessionMixnodes::<T>::insert(i as AuthorityIndex, mixnode);
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a mixnode for the following session.
+		#[pallet::call_index(0)]
+		#[pallet::weight(1)] // TODO
+		pub fn register(
+			origin: OriginFor<T>,
+			registration: RegistrationFor<T>,
+			_signature: AuthoritySignature,
+		) -> DispatchResult {
+			ensure_none(origin)?;
+
+			// Checked by ValidateUnsigned
+			debug_assert_eq!(registration.session_index, T::ValidatorSet::session_index());
+			debug_assert!(
+				Self::session_phase(frame_system::Pallet::<T>::block_number()).0 ==
+					SessionPhase::Register
+			);
+			debug_assert!(registration.authority_index < T::MaxAuthorities::get());
+
+			// Note we are registering for the _following_ session, so the if appears to be
+			// backwards...
+			if (registration.session_index & 1) == 0 {
+				OddSessionMixnodes::<T>::insert(registration.authority_index, registration.mixnode);
+			} else {
+				EvenSessionMixnodes::<T>::insert(
+					registration.authority_index,
+					registration.mixnode,
+				);
+			}
+
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			let Self::Call::register { registration, signature } = call else {
+				return InvalidTransaction::Call.into()
+			};
+
+			// Check session index matches
+			match registration.session_index.cmp(&T::ValidatorSet::session_index()) {
+				Ordering::Greater => return InvalidTransaction::Future.into(),
+				Ordering::Less => return InvalidTransaction::Stale.into(),
+				Ordering::Equal => (),
+			}
+
+			// Check registrations are open
+			let block_number = frame_system::Pallet::<T>::block_number();
+			if Self::session_phase(block_number).0 != SessionPhase::Register {
+				return InvalidTransaction::Future.into()
+			}
+
+			// Check authority index is valid
+			if registration.authority_index >= T::MaxAuthorities::get() {
+				return InvalidTransaction::BadProof.into()
+			}
+			let Some(authority_id) = NextAuthorityIds::<T>::get(registration.authority_index) else {
+				return InvalidTransaction::BadProof.into()
+			};
+
+			// Check the authority hasn't registered yet
+			if Self::already_registered(registration.session_index, registration.authority_index) {
+				return InvalidTransaction::Stale.into()
+			}
+
+			// Check signature
+			let signature_ok = registration.using_encoded(|encoded_registration| {
+				authority_id.verify(&encoded_registration, signature)
```

Comment:
Unsigned transactions are used so that validators don't have to pay a fee to register. To avoid spam, each validator can only register once per session. This is very similar to how the `im-online` pallet works.

Reply:
I don't know if `im-online` work this way I can probably same im-online transaction with no consequences.  Can say it is a bit out of my expertise area though.